src/golang/lib/engine/authenticate.go

```diff
@@ -43,11 +49,69 @@ func AuthenticateLambdaConfig(ctx context.Context, authConf auth.Config) error {
 		lambda_utils.SnowflakeConnectorType,
 	}
 
-	for _, functionType := range functionsToShip {
-		err := lambda_utils.CreateLambdaFunction(functionType, lambdaConf.RoleArn)
-		if err != nil {
-			return errors.Wrap(err, "Unable to Create Lambda Function")
-		}
+	// Run authentication only once since the credentials will be recorded.
+	err = lambda_utils.AuthenticateDockerToECR()
+	if err != nil {
+		return errors.Wrap(err, "Unable to authenticate Lambda Function.")
+	}
+
+	errGroup, errGroupCtx := errgroup.WithContext(ctx)
+	// Pull images on a concurrency of "MaxConcurrentDownload".
+	pullImageChannel := make(chan lambda_utils.LambdaFunctionType, len(functionsToShip))
+	defer close(pullImageChannel)
+	pushImageChannel := make(chan lambda_utils.LambdaFunctionType, len(functionsToShip))
+	defer close(pushImageChannel)
+	lambda_utils.AddFunctionTypeToChannel(functionsToShip[:], pullImageChannel)
+	for i := 0; i < MaxConcurrentDownload; i++ {
+		errGroup.Go(func() error {
+			for {
+				select {
+				case functionType := <-pullImageChannel:
+					lambdaFunctionType := functionType
+					err := lambda_utils.PullImageFromECR(lambdaFunctionType)
+					if err != nil {
+						return err
+					}
+					pushImageChannel <- functionType
+				case <-errGroupCtx.Done():
+					return errGroupCtx.Err()
+				default:
+					return nil
+				}
+			}
+		})
+	}
+
+	// Create a signal channel to flag when all the lambda functions have been created.
+	signalChannel := make(chan lambda_utils.LambdaFunctionType, len(functionsToShip))
```

Comment:
nit: Maybe a name like `outstandingWorkChannel` or `incompleteWorkChannel` might be a little more explanatory? I'd like to capture the fact that the channel is populated, but gradually drawn down to zero. Signal channel makes it sound like we're adding to the channel as the signaling mechanism.

Reply:
Also, I think it's important to note here in a comment that the contents of this channel do not matter, only the number of elements in it :)